timers/timers.go

```diff
@@ -0,0 +1,255 @@
+// Package timers provides a timer facility similar to the standard library's.
+// It's designed to be faster, but less accurate.
+package timers
+
+import (
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/andres-erbsen/clock"
+	"github.com/uber-go/atomic"
+)
+
+const cacheline = 64
+
+const (
+	// State machine for timers.
+	stateScheduled = iota
+	stateExpired
+	stateCanceled
+)
+
+// A Timer is a handle to a deferred operation.
+type Timer struct {
+	f        func()
+	deadline uint64 // in ticks
+	state    atomic.Int32
+	next     *Timer
+	tail     *Timer // only set on head
+}
+
+// Stop cancels the deferred operation. It returns whether or not the
+// cancellation succeeded.
+func (t *Timer) Stop() bool {
+	if t == nil {
+		return true // canceling a nil timer always succeeds
+	}
+	return t.state.CAS(stateScheduled, stateCanceled)
+}
+
+func (t *Timer) expire() bool {
+	return t.state.CAS(stateScheduled, stateExpired)
+}
+
+func (t *Timer) pushOne(head *Timer) *Timer {
+	head.next, head.tail = nil, nil
+	return t.push(head)
+}
+
+func (t *Timer) push(head *Timer) *Timer {
+	// link lists
+	if head == nil {
+		return t
+	}
+	if head.tail == nil {
```

Comment:
I may be misunderstanding, but I think that's only sort of the case; making `push` simpler just moves complexity to `gatherBuckets`. When we're adding any unexpired timers back into the bucket, we'd need to walk a bit of the list to see if it's a single element or >1 element, using `push` or `pushOne` as appropriate. To avoid that, we could hold the lock on the bucket for the entire method, rather than scoping it as narrowly as it is now.